app/models/vim_performance_daily.rb

```diff
@@ -33,11 +33,7 @@ def self.find_entries(ext_options)
   end
 
   def self.process_hourly_for_one_day(recs, options = {})
-    return [] if recs.blank?
```

Comment:
@chessbyte If `recs` is an association (as it is here), then yes. /cc @Fryguy @jrafanie 

1. It looks like `empty?` is now using `select 1 from table limit 1`.
4. Calling `each` right after `blank?` is no major performance hit - you would have run the query anyway.
5. `present?`/`blank?` is preferred since `nil.blank?` works. If you are going to process those records, then no major performance hit, but if you making a decision to create or something, then a `blank?` is very bad for GC, SQL, network, ...
6. `empty?` has a smaller query, but if you are then doing an `each` you end up querying the data twice. Also of note, `nil.empty?` blows up. So this tends to be avoided.
7. the subtlety just has me saying, avoid calling `blank? or `empty?` before an `each` block. It tends to lead to simpler and more performant code. (yes, there are cases when the prep for a query is expensive. but that is not very common)
2. Our reports/screens and rbac uses `requires().includes()`, so take a look at the second set of queries. It is currently more likely.
3. The second `blank?` took 0.5 seconds to create 3000 Vm objects.

---

```ruby
irb(main):056:0> vms = ExtManagementSystem.first.vms_and_templates ; nil
  ExtManagementSystem Load (0.5ms)  SELECT  "ext_management_systems".* FROM "ext_management_systems" ORDER BY "ext_management_systems"."id" ASC LIMIT $1  [["LIMIT", 1]]
  ExtManagementSystem Inst Including Associations (0.3ms - 1rows)
=> nil
irb(main):057:0> vms.empty?
  VmOrTemplate Exists (0.5ms)  SELECT  1 AS one FROM "vms" WHERE "vms"."ems_id" = $1 LIMIT $2  [["ems_id", 1], ["LIMIT", 1]]
=> false
irb(main):058:0> vms.empty?
  VmOrTemplate Exists (0.5ms)  SELECT  1 AS one FROM "vms" WHERE "vms"."ems_id" = $1 LIMIT $2  [["ems_id", 1], ["LIMIT", 1]]
=> false
irb(main):061:0> vms.blank?
  VmOrTemplate Load (17.3ms)  SELECT "vms".* FROM "vms" WHERE "vms"."ems_id" = $1  [["ems_id", 1]]
  VmOrTemplate Inst Including Associations (214.5ms - 3000rows)
=> false
irb(main):062:0> vms.blank?
=> false
irb(main):063:0> vms.present? # uses association from blank?
=> true
irb(main):064:0> vms.empty?
=> false
irb(main):065:0> vms.map(&:name).count
=> 3000
```
---
using `includes().references()` makes the queries much more expensive

```ruby
irb(main):077:0> vms = ExtManagementSystem.first.vms_and_templates.includes(:host).references(:host) ; nil
  ExtManagementSystem Load (0.5ms)  SELECT  "ext_management_systems".* FROM "ext_management_systems" ORDER BY "ext_management_systems"."id" ASC LIMIT $1  [["LIMIT", 1]]
  ExtManagementSystem Inst Including Associations (0.5ms - 1rows)
=> nil
irb(main):078:0> vms.empty?
   (2.2ms)  SELECT COUNT(DISTINCT "vms"."id") FROM "vms" LEFT OUTER JOIN "hosts" ON "hosts"."id" = "vms"."host_id" WHERE "vms"."ems_id" = $1  [["ems_id", 1]]
=> false
irb(main):079:0> vms.empty?
   (2.1ms)  SELECT COUNT(DISTINCT "vms"."id") FROM "vms" LEFT OUTER JOIN "hosts" ON "hosts"."id" = "vms"."host_id" WHERE "vms"."ems_id" = $1  [["ems_id", 1]]
=> false
irb(main):080:0> vms.blank?
  SQL (25.7ms)  SELECT "vms"."id" AS t0_r0, "vms"."vendor" AS t0_r1, "vms"."format" AS t0_r2, "vms"."version" AS t0_r3, "vms"."name" AS t0_r4, "vms"."description" AS t0_r5, "vms"."location" AS t0_r6, "vms"."config_xml" AS t0_r7, "vms"."autostart" AS t0_r8, "vms"."host_id" AS t0_r9, "vms"."last_sync_on" AS t0_r10, "vms"."created_on" AS t0_r11, "vms"."updated_on" AS t0_r12, "vms"."storage_id" AS t0_r13, "vms"."guid" AS t0_r14, "vms"."ems_id" AS t0_r15, "vms"."last_scan_on" AS t0_r16, "vms"."last_scan_attempt_on" AS t0_r17, "vms"."uid_ems" AS t0_r18, "vms"."retires_on" AS t0_r19, "vms"."retired" AS t0_r20, "vms"."boot_time" AS t0_r21, "vms"."tools_status" AS t0_r22, "vms"."standby_action" AS t0_r23, "vms"."power_state" AS t0_r24, "vms"."state_changed_on" AS t0_r25, "vms"."previous_state" AS t0_r26, "vms"."connection_state" AS t0_r27, "vms"."last_perf_capture_on" AS t0_r28, "vms"."registered" AS t0_r29, "vms"."busy" AS t0_r30, "vms"."smart" AS t0_r31, "vms"."memory_reserve" AS t0_r32, "vms"."memory_reserve_expand" AS t0_r33, "vms"."memory_limit" AS t0_r34, "vms"."memory_shares" AS t0_r35, "vms"."memory_shares_level" AS t0_r36, "vms"."cpu_reserve" AS t0_r37, "vms"."cpu_reserve_expand" AS t0_r38, "vms"."cpu_limit" AS t0_r39, "vms"."cpu_shares" AS t0_r40, "vms"."cpu_shares_level" AS t0_r41, "vms"."cpu_affinity" AS t0_r42, "vms"."ems_created_on" AS t0_r43, "vms"."template" AS t0_r44, "vms"."evm_owner_id" AS t0_r45, "vms"."ems_ref_obj" AS t0_r46, "vms"."miq_group_id" AS t0_r47, "vms"."linked_clone" AS t0_r48, "vms"."fault_tolerance" AS t0_r49, "vms"."type" AS t0_r50, "vms"."ems_ref" AS t0_r51, "vms"."ems_cluster_id" AS t0_r52, "vms"."retirement_warn" AS t0_r53, "vms"."retirement_last_warn" AS t0_r54, "vms"."vnc_port" AS t0_r55, "vms"."flavor_id" AS t0_r56, "vms"."availability_zone_id" AS t0_r57, "vms"."cloud" AS t0_r58, "vms"."retirement_state" AS t0_r59, "vms"."cloud_network_id" AS t0_r60, "vms"."cloud_subnet_id" AS t0_r61, "vms"."cloud_tenant_id" AS t0_r62, "vms"."raw_power_state" AS t0_r63, "vms"."publicly_available" AS t0_r64, "vms"."orchestration_stack_id" AS t0_r65, "vms"."retirement_requester" AS t0_r66, "vms"."tenant_id" AS t0_r67, "vms"."resource_group_id" AS t0_r68, "hosts"."id" AS t1_r0, "hosts"."name" AS t1_r1, "hosts"."hostname" AS t1_r2, "hosts"."ipaddress" AS t1_r3, "hosts"."vmm_vendor" AS t1_r4, "hosts"."vmm_version" AS t1_r5, "hosts"."vmm_product" AS t1_r6, "hosts"."vmm_buildnumber" AS t1_r7, "hosts"."created_on" AS t1_r8, "hosts"."updated_on" AS t1_r9, "hosts"."guid" AS t1_r10, "hosts"."ems_id" AS t1_r11, "hosts"."user_assigned_os" AS t1_r12, "hosts"."power_state" AS t1_r13, "hosts"."smart" AS t1_r14, "hosts"."settings" AS t1_r15, "hosts"."last_perf_capture_on" AS t1_r16, "hosts"."uid_ems" AS t1_r17, "hosts"."connection_state" AS t1_r18, "hosts"."ssh_permit_root_login" AS t1_r19, "hosts"."ems_ref_obj" AS t1_r20, "hosts"."admin_disabled" AS t1_r21, "hosts"."service_tag" AS t1_r22, "hosts"."asset_tag" AS t1_r23, "hosts"."ipmi_address" AS t1_r24, "hosts"."mac_address" AS t1_r25, "hosts"."type" AS t1_r26, "hosts"."failover" AS t1_r27, "hosts"."ems_ref" AS t1_r28, "hosts"."hyperthreading" AS t1_r29, "hosts"."ems_cluster_id" AS t1_r30, "hosts"."next_available_vnc_port" AS t1_r31, "hosts"."hypervisor_hostname" AS t1_r32, "hosts"."availability_zone_id" AS t1_r33 FROM "vms" LEFT OUTER JOIN "hosts" ON "hosts"."id" = "vms"."host_id" WHERE "vms"."ems_id" = $1  [["ems_id", 1]]
  VmOrTemplate Inst Including Associations (442.8ms - 3000rows)
=> false
irb(main):081:0> vms.empty?
=> false
```